GC hotpath optims - stf and branch mispredicts

arunetm opened the pull request:
*Using intrinsics to reduce Store Forward costs in the hotpath.
*Re arranging branches to reduce branch mispredict costs in the hotpath.

arunetm commented:
Combined Octane impact of #3632 #3633 

![image](https://user-images.githubusercontent.com/13296674/29990357-84b9274e-8f2e-11e7-90d2-a6ba13c18d87.png)

arunetm commented:
@dotnet-bot  test OSX static_osx_osx_debug please

arunetm commented:
@LouisLaf @MikeHolman fyi and please review. Will await for the IEPerflab results before merging this change.

arunetm commented:
@dotnet-bot test OSX static_osx_osx_debug please

atulkatti commented on lib/Common/Memory/MarkContext.inl:
```diff
@@ -22,8 +22,12 @@ bool MarkContext::AddMarkedObject(void * objectAddress, size_t objectSize)
+    *(__m128i *)&markCandidate = _mm_set_epi64x(objectSize, (__int64)objectAddress);
+#else
+    markCandidate.obj = (void**)objectAddress;
     markCandidate.byteCount = objectSize;
```
Should the bytecount be set in the #if case as well?

leirocks replied:
I guess that intrinsic set both addr and size fields though xmm register

arunetm commented on lib/Common/Memory/MarkContext.inl:
```diff
@@ -22,8 +22,12 @@ bool MarkContext::AddMarkedObject(void * objectAddress, size_t objectSize)
+    *(__m128i *)&markCandidate = _mm_set_epi64x(objectSize, (__int64)objectAddress);
+#else
+    markCandidate.obj = (void**)objectAddress;
     markCandidate.byteCount = objectSize;
```
Yes. We have a load for the entire object(128bits) right after the 64bit stores of address and size.  This uneven size stores and load prevents the Stores from being forwarded to the Load
bypassing the caches. 
With the intrinsic the compiler will generate similar sized store and load (128bit) and helps to avoid the above issue and saves a few cycles.

arunetm commented:
Update: Performance checkpoint (IEPerflab) evaluation shows improvements to Speedometer and Octane with no significant regressions.

arunetm commented:
@leirocks @atulkatti Can someone help with signing off this PR in preparation for merge? 

Updates
-Added code comments
-Updated headers
-Squished to a single commit
